sdk/core/System.ClientModel/src/Pipeline/MessageLoggingPolicy.cs

```diff
@@ -0,0 +1,372 @@
+    /// <see cref="ClientPipeline"/>.
+    /// </summary>
+    public static MessageLoggingPolicy Default { get; } = new MessageLoggingPolicy();
+
+    private readonly ClientLoggingOptions _loggingOptions;
+    private PipelineMessageLogger? _messageLogger;
+    private readonly string _clientAssembly = typeof(MessageLoggingPolicy).Assembly.GetName().Name!;
+
+    private bool _enableMessageContentLogging => _loggingOptions.EnableMessageContentLogging ?? ClientLoggingOptions.DefaultEnableMessageContentLogging;
+    private int _maxLength => _loggingOptions.MessageContentSizeLimit ?? ClientLoggingOptions.DefaultMessageContentSizeLimitBytes;
+
+    /// <summary>
+    /// Creates a new instance of the <see cref="MessageLoggingPolicy"/> class.
+    /// </summary>
+    /// <param name="options">The user-provided logging options object.</param>
+    public MessageLoggingPolicy(ClientLoggingOptions? options = default)
+    {
+        _loggingOptions = options ?? new();
+    }
+
+    /// <inheritdoc/>
+    public sealed override void Process(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex) =>
+        ProcessSyncOrAsync(message, pipeline, currentIndex, async: false).EnsureCompleted();
+
+    /// <inheritdoc/>
+    public sealed override async ValueTask ProcessAsync(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex) =>
+        await ProcessSyncOrAsync(message, pipeline, currentIndex, async: true).ConfigureAwait(false);
+
+    private async ValueTask ProcessSyncOrAsync(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex, bool async)
+    {
+        _messageLogger ??= new PipelineMessageLogger(_loggingOptions.GetPipelineMessageSanitizer(), _loggingOptions.LoggerFactory);
+
+        // EventLevel.Warning / LogLevel.Warning is the highest level logged by this policy.
+        // PipelineMessageLogger.IsEnabled checks to see: if an ILogger was provided - ensure it is enabled for at least LogLevel.Warning OR if
+        // an ILogger was not provided, see if an EventListener exists that is actively listening for at least EventLevel.Warning.
+        // If nothing is listening for logs then this policy immediately passes control to the next policy in the pipeline and returns when control
+        // is passed back. This avoids any performance hit from logging when no one is listening/collecting logs.
+        if (!_messageLogger.IsEnabled(LogLevel.Warning, EventLevel.Warning))
+        {
+            if (async)
+            {
+                await ProcessNextAsync(message, pipeline, currentIndex).ConfigureAwait(false);
+            }
+            else
+            {
+                ProcessNext(message, pipeline, currentIndex);
+            }
+            return;
+        }
+
+        PipelineRequest request = message.Request;
+
+        string requestId = message.Request.ClientRequestId ?? string.Empty;
```

Comment:
I think we should not invent new standard for it. https://www.w3.org/TR/trace-context/ is used for correlation and request identification and we should find a way to use it in the generic core.

Reply:
Yeah you're right. We will need a way to do back-compat with Azure.Core at some point, but I think trace context can be used as the default (which is all we need rn)